lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -839,6 +849,44 @@ void CombComponentOp::getAsmBlockArgumentNames(
 //===----------------------------------------------------------------------===//
 LogicalResult ControlOp::verify() { return verifyControlBody(*this); }
 
+// Helper function to set the return value and add elements to the queue.
+static void setOps(std::queue<Operation *> &que, SmallVector<InvokeOp, 4> &ret,
+                   Block *block) {
+  auto ops = block->getOps<InvokeOp>();
+  for (InvokeOp op : ops)
+    ret.push_back(op);
+  auto sIt = block->getOps<SeqOp>();
+  for (auto op : sIt)
+    que.push(op.getOperation());
+  auto pIt = block->getOps<ParOp>();
+  for (auto op : pIt)
+    que.push(op.getOperation());
+  auto wIt = block->getOps<WhileOp>();
+  for (auto op : wIt)
+    que.push(op.getOperation());
+  auto iIt = block->getOps<IfOp>();
+  for (auto op : iIt)
+    que.push(op.getOperation());
+}
+
+// Get the InvokeOp of the ControlOp.
+SmallVector<InvokeOp, 4> ControlOp::getInvokeOps() {
```

Comment:
Is it not sufficient just to do `getBodyBlock->getOps<InvokeOps>()` ? Does this not do a breadth search?

Reply:
Sorry, I don't seem to understand what you mean, a PR should not contain all the features?

Reply:
`just to do getBodyBlock->getOps<InvokeOps>()` is not correct, because the function of getBodyBlock->getOps in `while, if, par ,seq` is only to return the ops in this block, but some invoke operations may be hidden deeper and will not be detected.
for example
```
seq {  // Here getOps does not work, it can only take out a seq.
  seq {
      invoke.
  }
}

```

Reply:
> Sorry, I don't seem to understand what you mean, a PR should not contain all the features?

I mean that you're sending a large PR, and it makes it that much harder to review. 

https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#pull-request

> Consider splitting the large set of changes on small independent PRs that can be reviewed, tested and merged independently.

It's not required, but it will make reviewers' lives easier if you split this up into smaller PRs. Parts that don't make sense can be left with some failure message "Unimplemented".